File: y2023/kotlin/Day06.kt
package y2023

import general.Day
import general.product
import kotlin.math.sqrt

object Day06 : Day() {
    override val name = "Wait For It"

    override fun a1() {
        val lines = INPUT.readLines()
        val times = "\\d+".toRegex().findAll(lines[0]).map { it.value.toInt() }.toList()
        val dists = "\\d+".toRegex().findAll(lines[1]).map { it.value.toInt() }.toList()

        println(
            times.zip(dists)
                .map { (t, d) ->
//                    var sum = 0L
//                    for (ti in 1 until t) {
//                        val di = ti * (t - ti)
//                        if (di > d) sum++
//                    }
//                    sum
                    val det = t * t - 4 * d
                    if (det < 0) 0L
                    else {
                        val a = (t - sqrt(det.toDouble())) / 2
                        val b = (t + sqrt(det.toDouble())) / 2
                        b.toLong() - a.toLong()
                    }
                }.product()
        )
    }

    override fun a2() {
        val lines = INPUT.readLines()
        val time = lines[0].filter { it.isDigit() }.toLong()
        val dist = lines[1].filter { it.isDigit() }.toLong()

//        var sum = 0
//        for (ti in 1 until time) {
//            val di = ti * (time - ti)
//            if (di > dist) sum++
//        }
        val det = time * time - 4 * dist
        if (det < 0) println(0)
        else {
//            val a = (time - sqrt(det.toDouble())) / 2
//            val b = (time + sqrt(det.toDouble())) / 2
//            println(b.toLong() - a.toLong())
            println(sqrt(det.toDouble()).toLong())
        }
    }
}

File: y2023/haskell/day06.hs
import Data.Char (isDigit)
import Data.List (groupBy)
import Text.Regex.Base (getAllTextMatches)
import Text.Regex.TDFA ((=~))

main :: IO ()
main = do
    a1
    a2

a1 :: IO ()
a1 = do
    content <- readFile "y2023/inputs/d06"
    let times = read <$> getAllTextMatches (lines content !! 0 =~ "[[:digit:]]+") :: [Int]
    let dists = read <$> getAllTextMatches (lines content !! 1 =~ "[[:digit:]]+") :: [Int]
    print $ product [length [() | ti <- [1..t], ti * (t - ti) > d] | (t, d) <- zip times dists]


a2 :: IO ()
a2 = do
    content <- readFile "y2023/inputs/d06"
    let time = read (filter isDigit (lines content !! 0)) :: Int
    let dist = read (filter isDigit (lines content !! 1)) :: Int
    let det' = sqrt (fromIntegral $ time * time - 4 * dist)
    print $ floor det'
